### PR TITLE
[core] [next] [create-sitecore-jss] Passing configuration object to `defineConfig` in sitecore.config is now optional. 

### PR DESCRIPTION
**Pull request by yavorsk**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/content-sdk/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation

Main motivation of the PR is to simplify the starter app's sitecore.config:
- passing SitecoreConfigInput object to the `defineConfig` function is now optional.
- If nothing is provided to the `defineConfig` function, it will populate the required configuration properties from the environment variables defined in .env. Users can decide whether they want to override those default falback values via passing the SitecoreConfigInput object
- sitecore.config.example is added to the starter app to serve as an example
 
## Testing Details

- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts [hunk 1]
```diff
@@ -1,33 +1,6 @@
 import { defineConfig } from '@sitecore-content-sdk/nextjs/config';
 /**
  * @type {import('@sitecore-content-sdk/nextjs/config').SitecoreConfig}
+ * See documentation about defineConfig here https://doc.sitecore.com/xmc/en/developers/content-sdk/the-sitecore-configuration-file.html
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts]:
```suggestion
 * See the documentation for `defineConfig`:
 * https://doc.sitecore.com/xmc/en/developers/content-sdk/the-sitecore-configuration-file.html
```

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example
Comment (sc-hatimfarhathussain) [inline packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example:1]: would it be better the if naming of the example file should be something like this `sitecore.config.ts.example` , so users can identify it as a ts file. Just suggestion only.

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { defineConfig } from '@sitecore-content-sdk/nextjs/config';
+/**
+ * @type {import('@sitecore-content-sdk/nextjs/config').SitecoreConfig}
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example]:
```suggestion
 * @type {import('@sitecore-content-sdk/nextjs/config').SitecoreConfig}
 * See the documentation for `defineConfig`:
 * https://doc.sitecore.com/xmc/en/developers/content-sdk/the-sitecore-configuration-file.html
```

File: packages/core/src/config/models.ts
Comment (illiakovalenko) [inline packages/core/src/config/models.ts:1]: In some cases we know that specific environment variables can be used across of any framework
Can we mention how to control personalize timeouts and editing secret (anything else?) via env variables? tsdoc should mention that

File: packages/nextjs/src/config/define-config.ts [hunk 3]
```diff
@@ -4,6 +4,54 @@ import {
...
+export const getNextFallbackConfig = (config?: SitecoreConfigInput): SitecoreConfigInput => {
+  return {
+    ...config,
+    api: {
+      ...config?.api,
+      edge: {
+        ...config?.api?.edge,
+        contextId:
+          config?.api?.edge?.contextId ||
+          process.env.SITECORE_EDGE_CONTEXT_ID ||
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.ts]: Check for `SITECORE_EDGE_CONTEXT_ID` is not needed here, right? Since we should have this fallback in `core`

File: packages/nextjs/src/config/define-config.ts [hunk 4]
```diff
@@ -4,6 +4,54 @@ import {
...
+          process.env.NEXT_PUBLIC_SITECORE_EDGE_CONTEXT_ID ||
+          '',
+        clientContextId:
+          config?.api?.edge?.clientContextId || process.env.NEXT_PUBLIC_SITECORE_EDGE_CONTEXT_ID,
+        edgeUrl:
+          config?.api?.edge?.edgeUrl ||
+          process.env.SITECORE_EDGE_URL ||
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.ts]: Check for `SITECORE_EDGE_URL` is not needed here, right? Since we should have this fallback in `core`

**Review by illiakovalenko — CHANGES_REQUESTED** [review]

File: packages/nextjs/src/config/define-config.ts [hunk 5]
```diff
@@ -4,6 +4,54 @@ import {
...
+        ...config?.api?.local,
+        apiKey: config?.api?.local?.apiKey || process.env.NEXT_PUBLIC_SITECORE_API_KEY || '',
+        apiHost: config?.api?.local?.apiHost || process.env.NEXT_PUBLIC_SITECORE_API_HOST || '',
+      },
+    },
+    defaultSite: config?.defaultSite || process.env.NEXT_PUBLIC_SITECORE_SITE_NAME,
+    defaultLanguage: config?.defaultLanguage || process.env.NEXT_PUBLIC_DEFAULT_LANGUAGE,
+    multisite: {
+      ...config?.multisite,
+      enabled: config?.multisite?.enabled ?? true,
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.ts]: we should have a fallback to `true` in `core` package, right?

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example
Comment (yavorsk) [reply packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.example:1]: we have the `sitecore.config.ts` in the starter as well. but you are right, I will update :)

File: packages/nextjs/src/config/define-config.ts (hunk 3 above)
Comment (yavorsk) [reply packages/nextjs/src/config/define-config.ts]: hm yea we have the fallback in `core`. I left it in next as well to preserve the precedence of the two env variables, but I you are correct i guess it is not needed.

File: packages/nextjs/src/config/define-config.ts (hunk 5 above)
Comment (yavorsk) [reply packages/nextjs/src/config/define-config.ts]: fallback in core is `false`. that's why i included it in nextjs

File: packages/nextjs/src/config/define-config.test.ts [hunk 6]
```diff
@@ -64,4 +64,322 @@ describe('defineConfig', () => {
       });
     });
   });
+
+  describe('getNextFallbackConfig', () => {
```
Comment (art-alexeyenko) [inline packages/nextjs/src/config/define-config.test.ts]: instead of getNextFallbackConfig it's better to test defineConfig here - since this is what is used by end user.
Need to ensure that nextjs changes and core's `defineConfig` work well together, with everything that needs to be overriden, being overriden.

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts.example [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  defaultSite: process.env.NEXT_PUBLIC_SITECORE_SITE_NAME,
+  defaultLanguage: process.env.NEXT_PUBLIC_DEFAULT_LANGUAGE || 'en',
+  editingSecret: process.env.JSS_EDITING_SECRET,
+  redirects: {},
+  multisite: {
+    enabled: true,
+    useCookieResolution: () => process.env.VERCEL_ENV === 'preview',
+  },
+  personalize: {
```
Comment (art-alexeyenko) [inline packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts.example:33]: please add `enabled` to this and `redirects` section.
Overall I think we can go wild with the example and show all the different config options available, unless you agreed differently with Christian.

**Review by art-alexeyenko — CHANGES_REQUESTED** [review]

File: packages/nextjs/src/config/define-config.test.ts (hunk 6 above)
Comment (yavorsk) [reply packages/nextjs/src/config/define-config.test.ts]: hm but we are unittesting here, we should not rely on what's happening in core. Even if we use `defineConfig` in the tests we would need to mock core's `defineConfig` in that case. Unless I am missunderstanding your comment? 
And since all logic is happening in `getNextFallbackConfig` I decided to directly test that.

File: packages/core/src/config/models.ts
Comment (yavorsk) [reply packages/core/src/config/models.ts:1]: makes sense. i am wondering how to handle the general edge env vars, seems not accurate to mention them here.

Comment (illiakovalenko) [reply packages/core/src/config/models.ts:1]: yes, you can't do that

File: CHANGELOG.md [hunk 8]
```diff
@@ -16,6 +16,8 @@ Our versioning strategy is as follows:
 
 * `[core]` Introduce get locales functionality on content client ([#74](https://github.com/Sitecore/content-sdk/pull/74))([#75](https://github.com/Sitecore/content-sdk/pull/75))
 * `[cli]` Introduce "project" subcommands ([#73](https://github.com/Sitecore/content-sdk/pull/73))
+* `[core]``[next]``[create-sitecore-jss]` Passing configuration object to defineConfig in sitecore.config is now optional. Introduced sitecore.config.ts.example ([#90](https://github.com/Sitecore/content-sdk/pull/90))
```
Comment (illiakovalenko) [inline CHANGELOG.md]: If you will not add whitespace between backticks it will break preview

```suggestion
* `[core]` `[next]` `[create-sitecore-jss]` Passing configuration object to `defineConfig` in _sitecore.config_ is now optional. Introduced _sitecore.config.ts.example_ ([#90](https://github.com/Sitecore/content-sdk/pull/90))
```

File: packages/core/src/config/define-config.test.ts [hunk 9]
```diff
@@ -148,4 +149,42 @@ describe('define-config', () => {
       524,
     ]);
   });
+
+  describe('getFallbackConfig', () => {
```
Comment (illiakovalenko) [inline packages/core/src/config/define-config.test.ts:153]: There are tests above that are verifying `getFallbackConfig` as well. Can it be consolidated?

File: packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts.example [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  redirects: {
+    enabled: true,
+    locales: ['en'],
+  },
+  multisite: {
+    enabled: true,
+    useCookieResolution: () => process.env.VERCEL_ENV === 'preview',
+  },
+  personalize: {
+    enabled: true,
```
Comment (illiakovalenko) [inline packages/create-sitecore-jss/src/templates/nextjs/sitecore.config.ts.example]: personalization shouldn't be enabled by default in all the environments, since during development it should be disabled, so let's stick to our old approach and don't set it explicitly

File: packages/nextjs/src/config/define-config.ts [hunk 11]
```diff
@@ -4,6 +4,48 @@ import {
   SitecoreConfigInput as SitecoreConfigInputCore,
 } from '@sitecore-content-sdk/core/config';
 
+/**
+ * Provides default NextJs initial values from env variables for SitecoreConfig
+ * @returns default nextjs input config
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.ts]: tsdoc warning, please add optional `config` param here

File: packages/nextjs/src/config/define-config.ts (hunk 5 above)
Comment (illiakovalenko) [reply packages/nextjs/src/config/define-config.ts]: nope, this is not true, in core it fallsback to `true`:
https://github.com/Sitecore/content-sdk/blob/feature/JSS-5427/packages/core/src/config/models.ts#L121-L124

File: packages/nextjs/src/config/define-config.test.ts (hunk 6 above)
Comment (illiakovalenko) [reply packages/nextjs/src/config/define-config.test.ts]: yes, we don't need to re-test `defineConfig` from `core`
but I understand what exactly Artem would like to apply
you could mock `core` `defineConfig` and instead of executing and testing `getNextFallbackConfig` you could run nextjs `defineConfig`
Can you apply that? shouldn't really affect much

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
Looks really great! See my new comments, it should be a final round

File: packages/nextjs/src/config/define-config.ts [hunk 12]
```diff
@@ -4,6 +4,48 @@ import {
...
+    ...config,
+    api: {
+      ...config?.api,
+      edge: {
+        ...config?.api?.edge,
+        contextId:
+          config?.api?.edge?.contextId || process.env.NEXT_PUBLIC_SITECORE_EDGE_CONTEXT_ID || '',
+        clientContextId:
+          config?.api?.edge?.clientContextId || process.env.NEXT_PUBLIC_SITECORE_EDGE_CONTEXT_ID,
+        edgeUrl: config?.api?.edge?.edgeUrl || process.env.NEXT_PUBLIC_SITECORE_EDGE_URL,
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.ts:23]: please, add unit tests to cover `edgeUrl`

File: packages/nextjs/src/config/define-config.test.ts [hunk 13]
```diff
@@ -1,27 +1,374 @@
...
+      defineConfigModule.defineConfig(defaultConfig());
+      const resultConfig = defineConfigCoreStub.getCalls()[0].args[0];
+      expect(resultConfig.multisite?.enabled).to.be.undefined;
+    });
+
+    it('should be able to override default value of enabled', () => {
+      defineConfigModule.defineConfig({ ...defaultConfig(), multisite: { enabled: false } });
+      const resultConfig = defineConfigCoreStub.getCalls()[0].args[0];
+      expect(resultConfig.multisite?.enabled).to.be.false;
+    });
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.test.ts]: this is related to `enabled` property, right?
let's group it under `describe('enabled')`

File: packages/nextjs/src/config/define-config.test.ts [hunk 14]
```diff
@@ -1,27 +1,374 @@
...
+      });
+
+      it('should use the value from the config', () => {
+        defineConfigModule.defineConfig({
+          ...defaultConfig(),
+          personalize: { scope: 'custom-scope' },
+        });
+        const resultConfig = defineConfigCoreStub.getCalls()[0].args[0];
+        expect(resultConfig.personalize?.scope).to.equal('custom-scope');
+      });
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.test.ts]: these tests are specific to `scope`, let's group it under `describe('scope')`

File: packages/nextjs/src/config/define-config.test.ts [hunk 15]
```diff
@@ -1,27 +1,374 @@
...
+
+      it('should use the env var if config value not present', () => {
+        defineConfigModule.defineConfig(defaultConfig());
+        const resultConfig = defineConfigCoreStub.getCalls()[0].args[0];
+        expect(resultConfig.personalize?.scope).to.equal('custom-env-scope');
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.test.ts]: same as above, let's group
since if new properties will appear it will be required to restructure these tests

File: packages/nextjs/src/config/define-config.test.ts [hunk 16]
```diff
@@ -1,27 +1,374 @@
...
-        expect(config.disableStaticPaths).to.equal(false);
+        defineConfigModule.defineConfig(defaultConfig());
+        const resultConfig = defineConfigCoreStub.getCalls()[0].args[0];
+        expect(resultConfig.disableStaticPaths).to.equal(false);
       });
 
       it('should use the value from the config', () => {
-        const config = defineConfig({
+        defineConfigModule.defineConfig({
           disableStaticPaths: false,
```
Comment (illiakovalenko) [inline packages/nextjs/src/config/define-config.test.ts]: let's set it to `true` so we will test this edge case
because, currently `false` is our default value as well

**Review by illiakovalenko — APPROVED** [review]
Looks good to me 👍
I left few additional comments, but I approve in advance
